soccer/planning/RRTPlanner.cpp

```diff
@@ -1,52 +1,136 @@
+    if (prevPath->hit(*obstacles, hitTime, timeIntoPath)) {
+        return true;
+    }
+
+    // if the destination of the current path is greater than X m away
+    // from the target destination, we invalidate the path. This
+    // situation could arise if the path destination changed.
+    float goalPosDiff = (prevPath->destination()->pos - goal.pos).mag();
+    float goalVelDiff = (prevPath->destination()->vel - goal.vel).mag();
+    if (goalPosDiff > goalChangeThreshold() ||
+        goalVelDiff > goalChangeThreshold()) {
+        // FIXME: goalChangeThreshold shouldn't be used for velocities as it
+        // is above
+        return true;
+    }
+
+    return false;
+}
+
 std::unique_ptr<Path> RRTPlanner::run(
-    MotionInstant startInstant, MotionInstant endInstant,
+    MotionInstant start, MotionCommand cmd,
     const MotionConstraints& motionConstraints,
-    const Geometry2d::CompositeShape* obstacles) {
-    Planning::InterpolatedPath* path = new Planning::InterpolatedPath();
-    Geometry2d::Point goal = endInstant.pos;
-    _motionConstraints = motionConstraints;
-    vi = startInstant.vel;
-    vf = endInstant.vel;
+    const Geometry2d::ShapeSet* obstacles, std::unique_ptr<Path> prevPath) {
+    // This planner only works with commands of type 'PathTarget'
+    assert(cmd.getCommandType() == Planning::MotionCommand::PathTarget);
 
-    _obstacles = obstacles;
+    MotionInstant goal = cmd.getPlanningTarget();
 
     // Simple case: no path
-    if (startInstant.pos == goal) {
-        path->times.push_back(0);
-        path->waypoints.emplace_back(startInstant.pos, Geometry2d::Point());
+    if (start.pos == goal.pos) {
+        InterpolatedPath* path = new InterpolatedPath();
+        path->setStartTime(timestamp());
+        path->waypoints.emplace_back(
+            MotionInstant(start.pos, Geometry2d::Point()), 0);
         return unique_ptr<Path>(path);
     }
 
-    /// Locate a non blocked goal point
-    Geometry2d::Point newGoal = goal;
+    // Locate a goal point that is obstacle-free
+    boost::optional<Geometry2d::Point> prevGoal;
+    if (prevPath) prevGoal = prevPath->destination()->pos;
+    goal.pos = findNonBlockedGoal(goal.pos, prevGoal, obstacles);
+
+    // Replan if needed, otherwise return the previous path unmodified
+    if (shouldReplan(start, goal, motionConstraints, obstacles,
+                     prevPath.get())) {
+        // Run bi-directional RRT to generate a path.
+        InterpolatedPath* path =
+            runRRT(start, goal, motionConstraints, obstacles);
+
+        // If RRT failed, the path will be empty, so we need to add a single
+        // point to make it valid.
+        if (path && path->waypoints.empty()) {
+            path->waypoints.emplace_back(
+                MotionInstant(start.pos, Geometry2d::Point()), 0);
+        }
+        return unique_ptr<Path>(path);
+    } else {
+        return prevPath;
+    }
+}
 
+Geometry2d::Point RRTPlanner::findNonBlockedGoal(
```

Comment:
Not sure what you mean. This is used when the requested move target is inside an obstacle.  In that case, we try to find the closest non-blocked point to the one requested.

Handling a starting position that's inside an obstacle is a separate case.  The RRT handles this already at each step.  See `FixedStepTree.extend()`.